ARCHITECTURE.md

```diff
@@ -0,0 +1,197 @@
+path. They are lazily downloaded and during tests runtime using
+[MavenArtifactFetcher](https://github.com/robolectric/robolectric/blob/master/plugins/maven-dependency-resolver/src/main/java/org/robolectric/internal/dependency/MavenArtifactFetcher.java).
+
+Although Robolectric supports shadowing for Android framework classes, it is
+also possible for users to perform Robolectric instrumentation for any package
+(with the exception of built in Java packages). This enables shadowing of
+arbitrary third-party code.
+
+# Shadows
+
+By default when an Android method is invoked during a Robolectric test, the real
+Android framework code is invoked. This is because a lot of Android framework
+classes are pure Java code (e.g the
+[Intent](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/core/java/android/content/Intent.java)
+class or the
+[org.json](https://cs.android.com/android/platform/superproject/main/+/main:libcore/json/src/main/java/org/json/)
+package) and that code can run on the JVM without any modifications needed.
+
+However, there are cases where Robolectric needs to intercept and replace
+Android method calls. This most commonly occurs when Android system service or
+native methods are invoked. To do this, Robolectric uses a system called Shadow
+classes.
+
+Shadow classes are Java classes that contain the replacement code of Android
+methods when they are invoked. Each shadow class is bound to specific Android
+classes and methods through annotations. There are currently hundreds of shadow
+classes that can be found
+[here](https://github.com/robolectric/robolectric/tree/master/shadows/framework/src/main/java/org/robolectric/shadows).
+
+Shadow classes may optionally contain public apis APIs can customize the
+behavior of the methods they are shadowing.
+
+Robolectric allows tests to specify custom shadows as well to provide user
+defined implementation for Android classes.
+
+## Shadow Packages and the Robolectric Annotation Processor 
+
+There are two categories of shadows: Robolectric’s built-in shadows that are
+aggregated using the [Robolectric Annotation Processor
+(RAP)](https://github.com/robolectric/robolectric/blob/master/processor/src/main/java/org/robolectric/annotation/processing/RobolectricProcessor.java),
+and custom shadows that are commonly specified using `@Config(shadows = …)`. RAP
+is configured to process all of the shadow files that exist in Robolectric’s
+code, such as [framework
+shadows](https://github.com/robolectric/robolectric/tree/master/shadows/framework).
+It processes all of the source files and generates a source file that implements
+the
+[ShadowProvider](https://github.com/robolectric/robolectric/blob/master/shadowapi/src/main/java/org/robolectric/internal/ShadowProvider.java)
+interface. During runtime, Robolectric will use ServiceLoader to detect all
+ShadowProvider classes and load them as built-in shadows by default.
```

Comment:
Im a bit confused here. The sentence above seems to imply that there's one ShadowProvider that is built from every shadow, but this sentence seems to imply there are multiple ShadowProviders.

Reply:
Fixed.